### PR TITLE
Something about writing.

### PR DESCRIPTION
**Pull request by zipper-meng**
1. Edit module `protos`
2. Add wal reader & wal writer
4. Modify file_manager::write_at
5. Add async_rt::try_wake
6. Replace dependence  `crc` with `crc32fast`
7. tskv::write now process WalTask in a future

### CODE REVIEW

File: tskv/src/kvcore.rs [hunk 1]
```diff
@@ -12,39 +29,87 @@ use once_cell::sync::OnceCell;
...
 
+    pub fn run_wal_job(&self, mut receiver: UnboundedReceiver<WalTask>) {
+        let fm = Arc::clone(&self.file_manager);
+        let wal_opt = self.options.wal.clone();
+        let f = async move {
+            let mut wal_manager = WalFileManager::new(fm, wal_opt);
+            while let Some(x) = receiver.next().await {
+                match x {
+                    WalTask::Write { rows, cb } => {
+                        let ret = wal_manager.write(&rows).await;
```
Comment (roseboy-liu) [inline tskv/src/kvcore.rs:102]: result allways ok ?

File: tskv/src/wal.rs [hunk 2]
```diff
@@ -1,101 +1,350 @@
...
 }
 
-impl WriteAheadLogManager {
-    pub fn new(file_manager: &'static FileManager, dir: String) -> Self {
+unsafe impl Send for WalFileManager {}
+unsafe impl Sync for WalFileManager {}
+
+impl WalFileManager {
+    pub fn new(file_manager: Arc<FileManager>, config: option::WalConfig) -> Self {
         let mut fs_options = Options::default();
```
Comment (roseboy-liu) [inline tskv/src/wal.rs:302]: why define fs_options?

File: tskv/src/wal.rs [hunk 3]
```diff
@@ -1,101 +1,350 @@
...
+
+    #[snafu(display("Error with IO task"))]
+    FailedWithIoTask,
+}
+
+pub type WalResult<T> = std::result::Result<T, WalError>;
+
+pub enum WalEntryType {
+    Write = 1,
+    Delete = 2,
```
Comment (roseboy-liu) [inline tskv/src/wal.rs:81]: how deal with delete table & delete series

File: tskv/src/wal.rs [hunk 4]
```diff
@@ -105,121 +354,235 @@ impl WriteAheadLogManager {
...
 
-    fn roll_wal_file(&mut self) -> Result<()> {
-        let writer = &self.current_file_writer;
-        match writer {
-            Some(w) if w.size > SEGMENT_SIZE => self.new_wal_file(),
-            None => self.new_wal_file(),
-            _ => Ok(()),
-        }
-    }
+    pub async fn write(&mut self, rows: &[u8]) -> WalResult<()> {
```
Comment (roseboy-liu) [inline tskv/src/wal.rs:409]: restruct these three functions

File: tskv/src/wal.rs [hunk 5]
```diff
@@ -105,121 +354,235 @@ impl WriteAheadLogManager {
...
+
+            // TODO codes below may produce "future cannot be sent between threads safely" in `kvcore.rs`."
+            // self.file_manager
+            //     .write_at(Arc::clone(&writer.file), writer.size, buf.as_mut_slice())
+            //     .await;
         }
+        writer.size += buf.len() as u64;
+        writer
+            .file
+            .sync_all(FileSync::Soft)
```
Comment (roseboy-liu) [inline tskv/src/wal.rs:433]: use wal option

**Review by roseboy-liu — COMMENTED** [review]
it would be better if you split it to more pr.

**Review by roseboy-liu — APPROVED** [review]

File: tskv/src/wal.rs [hunk 6]
```diff
@@ -1,101 +1,350 @@
...
 
-impl WriteAheadLogManager {
-    pub fn new(file_manager: &'static FileManager, dir: String) -> Self {
+unsafe impl Send for WalFileManager {}
+unsafe impl Sync for WalFileManager {}
+
+impl WalFileManager {
+    pub fn new(file_manager: Arc<FileManager>, config: option::WalConfig) -> Self {
         let mut fs_options = Options::default();
         let fs_options = fs_options
```
[the remark this answers is not in the document]
Comment (zipper-meng) [reply tskv/src/wal.rs:303]: a never-used variable, but ignored

File: tskv/src/wal.rs (hunk 3 above)
Comment (zipper-meng) [reply tskv/src/wal.rs:81]: - delete table ,or only some of series, is `Delete`
- delete them with time range, then `DeleteRange`

File: tskv/src/kvcore.rs (hunk 1 above)
Comment (zipper-meng) [reply tskv/src/kvcore.rs:102]: will be fixed on next commit

File: tskv/src/wal.rs [hunk 7]
```diff
@@ -1,101 +1,350 @@
...
+}
+
+impl From<models::ColumnKeysWithRange<'_>> for WalEntryBlock {
+    fn from(cols: models::ColumnKeysWithRange<'_>) -> Self {
+        Self::DeleteRange((&cols).into())
+    }
+}
+
+impl WalEntryBlock {
+    pub fn new_write(bytes: &[u8]) -> Self {
```
[the remark this answers is not in the document]
Comment (zipper-meng) [reply tskv/src/wal.rs:174]: will be fixed on next commit

File: tskv/src/wal.rs (hunk 5 above)
Comment (zipper-meng) [reply tskv/src/wal.rs:433]: will be fixed on next commit

File: tskv/src/kvcore.rs (hunk 1 above)
Comment (zipper-meng) [reply tskv/src/kvcore.rs]: fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)

File: tskv/src/wal.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (zipper-meng) [reply tskv/src/wal.rs]: fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)

File: tskv/src/wal.rs (hunk 3 above)
Comment (zipper-meng) [reply tskv/src/wal.rs:82]: - delete table, or some of series, use `Delete`,
- delete them with a time range, use `DeleteRange`

File: tskv/src/wal.rs (hunk 4 above)
Comment (zipper-meng) [reply tskv/src/wal.rs]: fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)

File: tskv/src/wal.rs (hunk 5 above)
Comment (zipper-meng) [reply tskv/src/wal.rs]: fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)

**Review by roseboy-liu — APPROVED** [review]

**Review by roseboy-liu — APPROVED** [review]

**Pull request merged**